src/backend/app/tools.py

```diff
@@ -61,19 +62,79 @@ def generate_prediction(eval_params):
         {h3.geo_to_h3(lat, lon, hex_resolution) for lat, lon in coordinates}
     )
 
-    hull = alphashape.alphashape(coordinates, 1)
-    hull_points = list(mapping(hull)["coordinates"])
+    # hull = alphashape.alphashape(coordinates, 1)
+    # hull_points = list(mapping(hull)["coordinates"])
 
     annotation_hexagon_ids = {
         "presence": prediction_hexagon_ids,
         "absence": list()
     }
 
     return dict(
-        coordinates=coordinates.tolist(),
-        pred_loc_combined=pred_loc_combined.tolist(),
-        hull_points=hull_points,
+        # coordinates=coordinates.tolist(),
+        # pred_loc_combined=pred_loc_combined.tolist(),
+        # hull_points=hull_points,
         prediction_hexagon_ids=prediction_hexagon_ids,
         # set prediction_hexagon_ids as the starting point for the annotation
         annotation_hexagon_ids=annotation_hexagon_ids,
     )
+
+
+def generate_prediction_scores(eval_params, lowest_threshold=0.0001):
+    preds, locs = get_prediction(eval_params)
+
+    # switch lats and longs
+    locs[:, [1, 0]] = locs[:, [0, 1]]
+
+    # combine coordinates and predictions
+    pred_loc_combined = np.column_stack((locs, preds))
+    pred_loc_combined = np.float_(pred_loc_combined)
+    pred_loc_combined = pred_loc_combined[pred_loc_combined[:, 2] >= lowest_threshold]
+    hex_resolution = eval_params['hex_resolution']
+
+    # if a more detailed HeatMap needed, use `pred_loc_combined` for that
+    scores_dict={}
+    for lat, lon, pred in pred_loc_combined:
+        try:
```

Comment:
here too it is better to simply check if such a key is in the dictionary, and if not, then add it with an empty list, something like this:

```python
    scores_dict = {}
    for lat, lng, pred in pred_loc_combined:
        hex_index = h3.geo_to_h3(lat, lng, hex_resolution)
        if hex_index not in scores_dict:
            scores_dict[hex_index] = []
        scores_dict[hex_index].appnd(pred)
```

but if we only need the maximum score, then why just store it, no need to create a list for this

```python
    scores_dict = {}
    for lat, lng, pred in pred_loc_combined:
        hex_index = h3.geo_to_h3(lat, lng, hex_resolution)
        scores_dict[hex_index] = max(scores_dict.get(hex_index, 0), pred)
```

Reply:
This is what I mean when I say I code like a mathematician :D. Changed to first solution! I take your point, but I use the dictionary later on in other functions. Will consider changing it later.